### PR TITLE
introduce: `IRecordableTrie` 

### PR DESCRIPTION
**Pull request by riemannulus**
Sliced PR for review https://github.com/planetarium/libplanet/pull/3378.

### CODE REVIEW

File: Libplanet.Action.Tests/Mocks/MockAccount.cs [hunk 1]
```diff
@@ -0,0 +1,121 @@
+using System.Numerics;
+using Libplanet.Action.State;
+using Libplanet.Crypto;
+using Libplanet.Store.Trie;
+using Libplanet.Types.Assets;
+using Libplanet.Types.Tx;
+
+namespace Libplanet.Action.Tests.Mocks
+{
+    public static class MockAccount
```
Comment (greymistcube) [inline Libplanet.Action.Tests/Mocks/MockAccount.cs:10]: Eh... I suppose this is something akin to `MockAccountState`. However, it isn't clear what this class does is as this neither inherits an interface nor is there a description for it. This makes it very prone to creating "bad tests". 😣

File: Libplanet.Action/ActionEvaluator.cs [hunk 2]
```diff
@@ -34,18 +36,23 @@ public class ActionEvaluator : IActionEvaluator
...
         /// the states for a provided <see cref="Address"/>.</param>
+        /// <param name="keyValueStore">
+        /// A <see cref="IKeyValueStore"/> implementation to use for storing the states.
+        /// </param>
         /// <param name="actionTypeLoader"> A <see cref="IActionLoader"/> implementation using
         /// action type lookup.</param>
         public ActionEvaluator(
             PolicyBlockActionGetter policyBlockActionGetter,
             IBlockChainStates blockChainStates,
+            IKeyValueStore keyValueStore,
```
Comment (greymistcube) [inline Libplanet.Action/ActionEvaluator.cs:47]: Rather icky. In any case, does this need to be introduced at this moment? Think this might be an artifact from splitting #3378. 🙄

File: Libplanet.Action/ActionEvaluator.cs [hunk 3]
```diff
@@ -93,19 +100,20 @@ public IReadOnlyList<IActionEvaluation> Evaluate(IPreEvaluationBlock block)
...
-                else
+                if (!(policyBlockAction is null))
                 {
-                    previousState = evaluations.Count > 0
-                        ? evaluations.Last().OutputState
-                        : previousState;
-                    return evaluations.Add(
-                        EvaluatePolicyBlockAction(block, previousState)
-                    );
+                     previousState = evaluations.Count > 0
```
Comment (greymistcube) [inline Libplanet.Action/ActionEvaluator.cs:105]: Tab space of 5? 🙃

File: Libplanet.Action.Tests/Mocks/MockAccount.cs (hunk 1 above)
Comment (riemannulus) [reply Libplanet.Action.Tests/Mocks/MockAccount.cs:10]: Maybe using `Account` not `MockAccount` is better than that. What about you?

File: Libplanet.Action/ActionEvaluator.cs (hunk 2 above)
Comment (riemannulus) [reply Libplanet.Action/ActionEvaluator.cs:47]: No, it is from #3378 as you know. Might be not perfectly split from the previous PR...

File: Libplanet.Action/ActionEvaluator.cs (hunk 3 above)
Comment (riemannulus) [reply Libplanet.Action/ActionEvaluator.cs]: https://github.com/planetarium/libplanet/pull/3392/commits/6bbed288b2f7566eb8a924a19c2963663be0d2d4 amended.

File: Libplanet.Action.Tests/Mocks/MockAccount.cs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        public static IAccount Empty { get; } =
+            new Account(default, new MerkleTrie(new MemoryKeyValueStore()));
+
+        public static IAccount SetBalance(
+            this IAccount account,
+            Address address,
+            FungibleAssetValue amount) =>
+            SetBalance(account, (address, amount.Currency), amount);
+
+        public static IAccount SetBalance(
```
Comment (greymistcube) [inline Libplanet.Action.Tests/Mocks/MockAccount.cs]: I think return `type` should be the same as the class not `IAccount`.

File: Libplanet.Action.Tests/Mocks/MockAccount.cs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        {
+            var context = CreateContext(account, pair.Address);
+            var currentAmount = account.GetBalance(pair.Address, pair.Currency);
+            var balanceDelta = currentAmount - amount;
+            return balanceDelta.Sign switch
+            {
+                0 => account,
+                > 0 => account.BurnAsset(context, pair.Address, balanceDelta),
+                < 0 => account.MintAsset(context, pair.Address, balanceDelta.Abs()),
+            };
```
Comment (greymistcube) [inline Libplanet.Action.Tests/Mocks/MockAccount.cs]: Similar as above. Fixture setting should be agnostic to logic as much as possible. This can throw an `Exception` from `IAccount`'s implementation.

File: Libplanet.Action.Tests/Mocks/MockAccount.cs (hunk 1 above)
Comment (greymistcube) [reply Libplanet.Action.Tests/Mocks/MockAccount.cs]: Modified documentation from `MockAccountState` would be somewhat helpful. 😥

File: Libplanet.Action.Tests/Mocks/MockAccount.cs (hunk 4 above)
Comment (riemannulus) [reply Libplanet.Action.Tests/Mocks/MockAccount.cs]: but this is some kind of extension method for `IAccount`. note `this IAccount account`.

File: Libplanet.Action.Tests/Mocks/MockAccount.cs (hunk 5 above)
Comment (riemannulus) [reply Libplanet.Action.Tests/Mocks/MockAccount.cs]: Hmm... but if the `IAccount` implementation was correct, it would be fine I guess...? Can you explain some kind of `Exception` thrown case if `IAccount.BurnAsset` and `IAccount.MintAsset` are implemented as spec?

File: Libplanet.Action.Tests/Mocks/MockAccount.cs (hunk 1 above)
Comment (riemannulus) [reply Libplanet.Action.Tests/Mocks/MockAccount.cs]: https://github.com/planetarium/libplanet/pull/3392/commits/a42078fb2141e9f2a31d74b82b6f806ca89aa013 from `MockAccountState`

File: Libplanet.Action.Tests/Mocks/MockAccount.cs (hunk 4 above)
Comment (greymistcube) [reply Libplanet.Action.Tests/Mocks/MockAccount.cs]: Yes, and I'm suggesting this shouldn't be an extension for an `IAccount` 🙃.
`IAccount`'s behavior is complicated enough as it is. It isn't at all clear how "setting a balance" is achieved.
As any kind of change to balance requires `IActionContext`, this hides implicit `IActionContext` creation. Also, if the currency in question has minters, I don't think there is any way to set balance to a non-minter `Address`.

Comment (riemannulus) [reply Libplanet.Action.Tests/Mocks/MockAccount.cs]: Okay, then use which implementation in `IUnRecordableTrie` in `MockAccount`? `new UnRecordableTrie(new MerkleTrie())`?

**Comment by riemannulus** [comment]
I'm overhauling and refactoring `IUnRecordableTrie`, could you check out one more time? @greymistcube

File: Libplanet/Blockchain/BlockChain.Evaluate.cs [hunk 6]
```diff
@@ -106,32 +107,40 @@ public partial class BlockChain
...
+                else
+                {
+                    _logger.Information(
+                        "Block #{BlockIndex} pre-evaluation hash {PreEvaluationHash} " +
+                        "did not produce any action evaluations",
                         block.Index,
                         block.PreEvaluationHash);
-
-                return rootHash;
+                    return Store.GetStateRootHash(block.PreviousHash) ?? default;
```
[the remark this answers is not in the document]
Comment (greymistcube) [reply Libplanet/Blockchain/BlockChain.Evaluate.cs:142]: Eh, then I guess it's the problem with `null`ables again. Either we need to change `Block.StateRootHash` to `null`able or `IStore.GetStateRootHash()` to return "empty state root hash" (`MerkleTrie.EmptyRootHash`). I'll make an issue.

File: Libplanet.Store/Trie/IRecordableTrie.cs [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+{
+    public interface IRecordableTrie : ITrie
+    {
+         /// <summary>
+         /// Cleans up and stores the <see cref="IRecordableTrie"/> in storage.
+         /// </summary>
+         /// <returns>Returns new committed <see cref="IRecordableTrie"/>.</returns>
+        IRecordableTrie Commit();
+
+        new IRecordableTrie Set(in KeyBytes key, IValue value);
```
Comment (greymistcube) [inline Libplanet.Store/Trie/IRecordableTrie.cs:13]: Why does `IRecordableTrie` need `Set()`? Shouldn't `Set()` be done via `ITrie`? 🤔

Comment (riemannulus) [reply Libplanet.Store/Trie/IRecordableTrie.cs:13]: https://github.com/planetarium/libplanet/pull/3392/commits/7cc6de819734f5c25355005807a5c31ccd52025b for convenience

Comment (greymistcube) [reply Libplanet.Store/Trie/IRecordableTrie.cs:13]: Weighing the convenience gained against unnecessary cluttering of an `interface`, if it was up to me, I'd say it'd be better to get rid of the whole extensions. 🙄

File: Libplanet.Action.Tests/Mocks/MockAccount.cs (hunk 1 above)
Comment (greymistcube) [inline Libplanet.Action.Tests/Mocks/MockAccount.cs:10]: Welp, leave I'll let it slide for now. We might need to get rid of this entirely or overhaul the whole thing.

File: Libplanet.Tests/Action/AccountTest.cs [hunk 8]
```diff
@@ -52,15 +51,14 @@ protected AccountTest(ITestOutputHelper output)
                 Currency.Capped("QUUX", 0, (100, 0), minter: _addr[0]),
             };
 
-            _initState = MockAccountState.Empty
+            _initAccount = new Account(default, new MerkleTrie(new MemoryKeyValueStore()))
                 .SetState(_addr[0], (Text)"a")
                 .SetState(_addr[1], (Text)"b")
                 .SetBalance(_addr[0], _currencies[0], 5)
-                .SetBalance(_addr[0], _currencies[1], -10)
+                .SetBalance(_addr[0], _currencies[1], 10)
```
Comment (greymistcube) [inline Libplanet.Tests/Action/AccountTest.cs:58]: Why this change?

File: Libplanet.Action/State/IAccount.cs [hunk 9]
```diff
@@ -47,6 +48,10 @@ public interface IAccount : IAccountState
         [Pure]
         IAccountDelta Delta { get; }
 
+        public ITrie Trie { get; }
+
+        public BlockHash BlockHash { get; }
```
Comment (greymistcube) [inline Libplanet.Action/State/IAccount.cs:53]: Why does `IAccount` need `BlockHash`? 🤔

**Review by greymistcube — APPROVED** [review]
Eh, let's merge this and deal with other issues (#3396, #3394) separately. Not sure why CI is acting up tho. 🙄
